client/auth/login.jsx

```diff
@@ -22,47 +22,79 @@ import Main from 'calypso/components/main';
 import Notice from 'calypso/components/notice';
 import SelfHostedInstructions from './self-hosted-instructions';
 import WordPressLogo from 'calypso/components/wordpress-logo';
-import { login } from 'calypso/lib/oauth-store/actions';
 import { recordGoogleEvent } from 'calypso/state/analytics/actions';
 import { localizeUrl } from 'calypso/lib/i18n-utils';
+import * as OAuthToken from 'calypso/lib/oauth-token';
+import { errors as errorTypes, makeAuthRequest, bumpStats } from './oauth-utils';
+
+const debug = debugFactory( 'calypso:oauth' );
 
 export class Auth extends Component {
 	state = {
 		login: '',
 		password: '',
 		auth_code: '',
-		...AuthStore.get(),
+		requires2fa: false,
+		inProgress: false,
+		errorLevel: false,
+		errorMessage: false,
 	};
 
-	componentDidMount() {
-		AuthStore.on( 'change', this.refreshData );
-	}
-
-	componentWillUnmount() {
-		AuthStore.off( 'change', this.refreshData );
-	}
-
 	getClickHandler = ( action ) => () =>
 		this.props.recordGoogleEvent( 'Me', 'Clicked on ' + action );
 
 	getFocusHandler = ( action ) => () =>
 		this.props.recordGoogleEvent( 'Me', 'Focused on ' + action );
 
-	refreshData = () => {
-		this.setState( AuthStore.get() );
-	};
-
 	focusInput = ( input ) => {
 		if ( this.state.requires2fa && this.state.inProgress === false ) {
 			input.focus();
 		}
 	};
 
-	submitForm = ( event ) => {
+	handleAuthError = ( error ) => {
+		const stateChanges = {
+			errorLevel: 'is-error',
+			requires2fa: false,
+			inProgress: false,
+			errorMessage: error.message,
+		};
+
+		debug( 'Error processing login: ' + stateChanges.errorMessage );
+
+		if ( error.type === errorTypes.ERROR_REQUIRES_2FA ) {
+			stateChanges.requires2fa = true;
+			stateChanges.errorLevel = 'is-info';
+		} else if ( error.type === errorTypes.ERROR_INVALID_OTP ) {
+			stateChanges.requires2fa = true;
+		}
+
+		return stateChanges;
+	};
+
+	handleLogin = ( response ) => {
+		debug( 'Access token received' );
+
+		OAuthToken.setToken( response.body.access_token );
+
+		document.location.replace( '/' );
+	};
+
+	submitForm = async ( event ) => {
 		event.preventDefault();
 		event.stopPropagation();
 
-		login( this.state.login, this.state.password, this.state.auth_code );
+		this.setState( { inProgress: true, errorLevel: false, errorMessage: false } );
+
+		try {
+			const { login, password, auth_code } = this.state;
+			const response = await makeAuthRequest( login, password, auth_code );
+
+			this.handleLogin( response );
+		} catch ( error ) {
+			bumpStats( error );
+			this.setState( this.handleAuthError( error ) );
```

Comment:
It would be nice if there was some symmetry in how success and failure are handled:
```js
try {
  const response = makeAuthRequest();
  this.handleAuthSuccess( response );
} catch ( error ) {
  this.handleAuthError( error );
}
```
I.e., hide the `bumpStats` and `setState` calls inside `handleAuthError`.

Reply:
addressed in bf08712 and 9299b01